### PR TITLE
Don't fail on cert_find in the UI on a CA-less installation

### PR DESCRIPTION
**Pull request by rcritten**
The UI is using cert_find to show certificates for users, services,
hosts, etc and in a CA-less environment there is no CA to query.

The side-effect is that no certificates at all will be shown in the
UI, even if they exist within the entry itself.

Fixes part of https://pagure.io/freeipa/issue/7202

Signed-off-by: Rob Crittenden <rcritten@redhat.com>

A separate change is going to be needed to perform a fallback and display any certificates stored within the entry itself.

This patch alone does not complete the ticket.

### CODE REVIEW

**Comment by stlaz** [comment]
This is wrong, you can still add a certificate to a user/service/host even though CA is not installed. You need to work around this better.

**Comment by stlaz** [comment]
Please rebase on current master, Travis is switching `pep8` to `pycodestyle`

**Comment by rcritten** [comment]
@stlaz I said exactly the same thing. This patch just makes it not crash. @pvomacka is going to need to fix the UI to otherwise display certificates when cert_find is not available.

**Comment by stlaz** [comment]
I tried to see what the behavior was in 4.4.* but as our lab lies in ashes, I only have to guess. If I remember the code back there correctly, it was still possible to issue `cert-find` back then, with some restrictions. I would not like to treat this regression by introducing another regression, could you maybe see 4.4.* code and implement the cert-find dependency on FreeIPA CA in the same way there was?

**Comment by rcritten** [comment]
I'll see what 4.4.x did but either way this is a bug in master. Install a CA-less master and run cert-find and watch it fail.

**Comment by rcritten** [comment]
Ok, it is surprising to me that cert_find works at all in CA-less, it wasn't originally designed to.

In any case this was broken in commit eb6d4c3037d0cc269a7924745f1cbd8f647e6e1a

It is easy enough to fix the traceback by testing for api.env.enable_ra but the details of the certificates don't show on the CLI because the lookup is deferred (it works in the UI as-is). I'll keep digging.

File: ipaserver/plugins/cert.py [hunk 1]
```diff
@@ -1580,6 +1585,8 @@ def _ldap_search(self, all, pkey_only, no_members, **options):
 
     def execute(self, criteria=None, all=False, raw=False, pkey_only=False,
                 no_members=True, timelimit=None, sizelimit=None, **options):
+        self.ca_enabled = self.api.Command.ca_is_enabled()['result']
```
Comment (stlaz) [inline ipaserver/plugins/cert.py]: `ca_enabled` seems to be locked in Python 2:
```
[Mon Oct 30 16:23:02.474784 2017] [wsgi:error] [pid 4859] [remote 172.17.0.2:47078] ipa: ERROR: non-public: AttributeError: locked: cannot set cert_find.ca_enabled to True
[Mon Oct 30 16:23:02.474810 2017] [wsgi:error] [pid 4859] [remote 172.17.0.2:47078] Traceback (most recent call last):
[Mon Oct 30 16:23:02.474813 2017] [wsgi:error] [pid 4859] [remote 172.17.0.2:47078]   File "/usr/lib/python2.7/site-packages/ipaserver/rpcserver.py", line 370, in wsgi_execute
[Mon Oct 30 16:23:02.474815 2017] [wsgi:error] [pid 4859] [remote 172.17.0.2:47078]     result = command(*args, **options)
[Mon Oct 30 16:23:02.474817 2017] [wsgi:error] [pid 4859] [remote 172.17.0.2:47078]   File "/usr/lib/python2.7/site-packages/ipalib/frontend.py", line 450, in __call__
[Mon Oct 30 16:23:02.474820 2017] [wsgi:error] [pid 4859] [remote 172.17.0.2:47078]     return self.__do_call(*args, **options)
[Mon Oct 30 16:23:02.474822 2017] [wsgi:error] [pid 4859] [remote 172.17.0.2:47078]   File "/usr/lib/python2.7/site-packages/ipalib/frontend.py", line 478, in __do_call
[Mon Oct 30 16:23:02.474824 2017] [wsgi:error] [pid 4859] [remote 172.17.0.2:47078]     ret = self.run(*args, **options)
[Mon Oct 30 16:23:02.474826 2017] [wsgi:error] [pid 4859] [remote 172.17.0.2:47078]   File "/usr/lib/python2.7/site-packages/ipalib/frontend.py", line 800, in run
[Mon Oct 30 16:23:02.474829 2017] [wsgi:error] [pid 4859] [remote 172.17.0.2:47078]     return self.execute(*args, **options)
[Mon Oct 30 16:23:02.474831 2017] [wsgi:error] [pid 4859] [remote 172.17.0.2:47078]   File "/usr/lib/python2.7/site-packages/ipaserver/plugins/baseldap.py", line 1581, in execute
[Mon Oct 30 16:23:02.474833 2017] [wsgi:error] [pid 4859] [remote 172.17.0.2:47078]     delete_entry(pkey)
[Mon Oct 30 16:23:02.474835 2017] [wsgi:error] [pid 4859] [remote 172.17.0.2:47078]   File "/usr/lib/python2.7/site-packages/ipaserver/plugins/baseldap.py", line 1534, in delete_entry
[Mon Oct 30 16:23:02.474837 2017] [wsgi:error] [pid 4859] [remote 172.17.0.2:47078]     dn = callback(self, ldap, dn, *nkeys, **options)
[Mon Oct 30 16:23:02.474840 2017] [wsgi:error] [pid 4859] [remote 172.17.0.2:47078]   File "/usr/lib/python2.7/site-packages/ipaserver/plugins/host.py", line 842, in pre_callback
[Mon Oct 30 16:23:02.474842 2017] [wsgi:error] [pid 4859] [remote 172.17.0.2:47078]     certs = self.api.Command.cert_find(host=keys)['result']
[Mon Oct 30 16:23:02.474852 2017] [wsgi:error] [pid 4859] [remote 172.17.0.2:47078]   File "/usr/lib/python2.7/site-packages/ipalib/frontend.py", line 450, in __call__
[Mon Oct 30 16:23:02.474854 2017] [wsgi:error] [pid 4859] [remote 172.17.0.2:47078]     return self.__do_call(*args, **options)
[Mon Oct 30 16:23:02.474856 2017] [wsgi:error] [pid 4859] [remote 172.17.0.2:47078]   File "/usr/lib/python2.7/site-packages/ipalib/frontend.py", line 478, in __do_call
[Mon Oct 30 16:23:02.474859 2017] [wsgi:error] [pid 4859] [remote 172.17.0.2:47078]     ret = self.run(*args, **options)
[Mon Oct 30 16:23:02.474861 2017] [wsgi:error] [pid 4859] [remote 172.17.0.2:47078]   File "/usr/lib/python2.7/site-packages/ipalib/frontend.py", line 800, in run
[Mon Oct 30 16:23:02.474863 2017] [wsgi:error] [pid 4859] [remote 172.17.0.2:47078]     return self.execute(*args, **options)
[Mon Oct 30 16:23:02.474865 2017] [wsgi:error] [pid 4859] [remote 172.17.0.2:47078]   File "/usr/lib/python2.7/site-packages/ipaserver/plugins/cert.py", line 1588, in execute
[Mon Oct 30 16:23:02.474867 2017] [wsgi:error] [pid 4859] [remote 172.17.0.2:47078]     self.ca_enabled = self.api.Command.ca_is_enabled()['result']
[Mon Oct 30 16:23:02.474869 2017] [wsgi:error] [pid 4859] [remote 172.17.0.2:47078]   File "/usr/lib/python2.7/site-packages/ipalib/base.py", line 134, in __setattr__
[Mon Oct 30 16:23:02.474871 2017] [wsgi:error] [pid 4859] [remote 172.17.0.2:47078]     SET_ERROR % (self.__class__.__name__, name, value)
[Mon Oct 30 16:23:02.474873 2017] [wsgi:error] [pid 4859] [remote 172.17.0.2:47078] AttributeError: locked: cannot set cert_find.ca_enabled to True
```

Comment (rcritten) [reply ipaserver/plugins/cert.py]: Switched to using context instead. It is concerning that python3 allowed this though.

**Comment by stlaz** [comment]
Seems to work, solution is a bit hacky but it would seem that there's no other way. ACK.

**Comment by stlaz** [comment]
master:

* 32c64a78cef499c05e4ffa71d033692571297dca Fix cert-find for CA-less installations

**Comment by stlaz** [comment]
@rcritten please create a manual backport for `ipa-4-5`.

**Pull request closed without merging**